src/ol/render/canvas/TextBuilder.js

```diff
@@ -427,11 +431,14 @@ class CanvasTextBuilder extends CanvasBuilder {
         this.text_,
         this.textKey_,
         this.strokeKey_,
-        this.fillKey_,
+        this.fillKey_ ? defaultFillStyle : this.fillKey_,
         this.textOffsetX_,
         this.textOffsetY_,
         geometryWidths,
       ]);
+      if (textState.backgroundFill) {
```

Comment:
The change in fills state and the additional SET_FILL_STYLE instruction for hit detection would be easier to read if added right above line 411. This would also make a nice symmetry with the reset here.